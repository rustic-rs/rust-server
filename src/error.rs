//! Error types

use abscissa_core::error::{BoxError, Context};
use axum::http::StatusCode;
use axum::response::{IntoResponse, Response};
use std::{
    fmt::{self, Display},
    io,
    ops::Deref,
    result::Result,
};

pub type AppResult<T> = Result<T, Error>;
pub type ApiResult<T> = Result<T, ApiErrorKind>;

/// Kinds of errors
#[derive(Clone, Debug, Eq, thiserror::Error, PartialEq)]
pub enum ErrorKind {
    /// Error in configuration file
    #[error("config error")]
    Config,

    /// Input/output error
    #[error("I/O error")]
    Io,

    /// General storage error
    #[error("storage error")]
    GeneralStorageError,

    /// Missing user input
    #[error("missing user input")]
    MissingUserInput,
}

#[derive(Debug, thiserror::Error, displaydoc::Display)]
pub enum ApiErrorKind {
    /// Internal server error: `{0}`
    InternalError(String),
    /// Bad request: `{0}`
    BadRequest(String),
    /// Filename `{0}` not allowed
    FilenameNotAllowed(String),
    /// Path `{0}` is ambiguous with internal types and not allowed
    AmbiguousPath(String),
    /// Path `{0}` not allowed
    PathNotAllowed(String),
    /// Path `{0}` is not valid
    InvalidPath(String),
    /// Path `{0}` is not valid unicode
    NonUnicodePath(String),
    /// Creating directory failed: `{0}`
    CreatingDirectoryFailed(String),
    /// Not yet implemented
    NotImplemented,
    /// File not found: `{0}`
    FileNotFound(String),
    /// Fetting file metadata failed: `{0}`
    GettingFileMetadataFailed(String),
    /// Range not valid
    RangeNotValid,
    /// Seeking file failed
    SeekingFileFailed,
    /// Multipart range not implemented
    MultipartRangeNotImplemented,
    /// General range error
    GeneralRange,
    /// Conversion from length to u64 failed
    ConversionToU64Failed,
    /// Opening file failed: `{0}`
    OpeningFileFailed(String),
    /// Writing file failed: `{0}`
    WritingToFileFailed(String),
    /// Finalizing file failed: `{0}`
    FinalizingFileFailed(String),
    /// Getting file handle failed
    GettingFileHandleFailed,
    /// Removing file failed: `{0}`
    RemovingFileFailed(String),
    /// Reading from stream failed
    ReadingFromStreamFailed,
    /// Removing repository folder failed: `{0}`
    RemovingRepositoryFailed(String),
    /// Bad authentication header
    AuthenticationHeaderError,
    /// Failed to authenticate user: `{0}`
    UserAuthenticationError(String),
    /// General Storage error: `{0}`
    GeneralStorageError(String),
}

impl IntoResponse for ApiErrorKind {
    fn into_response(self) -> Response {
        let response = match self {
            ApiErrorKind::InternalError(err) => (
                StatusCode::INTERNAL_SERVER_ERROR,
                format!("Internal server error: {}", err),
            ),
            ApiErrorKind::BadRequest(err) => (
                StatusCode::BAD_REQUEST,
                format!("Internal server error: {}", err),
            ),
            ApiErrorKind::FilenameNotAllowed(filename) => (
                StatusCode::FORBIDDEN,
                format!("filename {filename} not allowed"),
            ),
            ApiErrorKind::AmbiguousPath(path) => (
                StatusCode::FORBIDDEN,
                format!("path {path} is ambiguous with internal types and not allowed"),
            ),
            ApiErrorKind::PathNotAllowed(path) => {
                (StatusCode::FORBIDDEN, format!("path {path} not allowed"))
            }
            ApiErrorKind::NonUnicodePath(path) => (
                StatusCode::BAD_REQUEST,
                format!("path {path} is not valid unicode"),
            ),
            ApiErrorKind::InvalidPath(path) => {
                (StatusCode::BAD_REQUEST, format!("path {path} is not valid"))
            }
            ApiErrorKind::CreatingDirectoryFailed(err) => (
                StatusCode::INTERNAL_SERVER_ERROR,
                format!("error creating dir: {:?}", err),
            ),
            ApiErrorKind::NotImplemented => (
                StatusCode::NOT_IMPLEMENTED,
                "not yet implemented".to_string(),
            ),
            ApiErrorKind::FileNotFound(path) => {
                (StatusCode::NOT_FOUND, format!("file not found: {path}"))
            }
            ApiErrorKind::GettingFileMetadataFailed(err) => (
                StatusCode::INTERNAL_SERVER_ERROR,
                format!("error getting file metadata: {err}"),
            ),
            ApiErrorKind::RangeNotValid => (StatusCode::BAD_REQUEST, "range not valid".to_string()),
            ApiErrorKind::SeekingFileFailed => (
                StatusCode::INTERNAL_SERVER_ERROR,
                "error seeking file".to_string(),
            ),
            ApiErrorKind::MultipartRangeNotImplemented => (
                StatusCode::NOT_IMPLEMENTED,
                "multipart range not implemented".to_string(),
            ),
            ApiErrorKind::ConversionToU64Failed => (
                StatusCode::INTERNAL_SERVER_ERROR,
                "error converting length to u64".to_string(),
            ),
            ApiErrorKind::OpeningFileFailed(err) => (
                StatusCode::INTERNAL_SERVER_ERROR,
                format!("error opening file: {err}"),
            ),
            ApiErrorKind::WritingToFileFailed(err) => (
                StatusCode::INTERNAL_SERVER_ERROR,
                format!("error writing file: {err}"),
            ),
            ApiErrorKind::FinalizingFileFailed(err) => (
                StatusCode::INTERNAL_SERVER_ERROR,
                format!("error finalizing file: {err}"),
            ),
            ApiErrorKind::GettingFileHandleFailed => (
                StatusCode::INTERNAL_SERVER_ERROR,
                "error getting file handle".to_string(),
            ),
            ApiErrorKind::RemovingFileFailed(err) => (
                StatusCode::INTERNAL_SERVER_ERROR,
                format!("error removing file: {err}"),
            ),
            ApiErrorKind::GeneralRange => {
                (StatusCode::INTERNAL_SERVER_ERROR, "range error".to_string())
            }
            ApiErrorKind::ReadingFromStreamFailed => (
                StatusCode::INTERNAL_SERVER_ERROR,
                "error reading from stream".to_string(),
            ),
            ApiErrorKind::RemovingRepositoryFailed(err) => (
                StatusCode::INTERNAL_SERVER_ERROR,
                format!("error removing repository folder: {:?}", err),
            ),
            ApiErrorKind::AuthenticationHeaderError => (
                StatusCode::FORBIDDEN,
                "Bad authentication header".to_string(),
            ),
            ApiErrorKind::UserAuthenticationError(err) => (
                StatusCode::FORBIDDEN,
                format!("Failed to authenticate user: {:?}", err),
            ),
            ApiErrorKind::GeneralStorageError(err) => (
                StatusCode::INTERNAL_SERVER_ERROR,
                format!("Storage error: {:?}", err),
            ),
        };

        response.into_response()
    }
}

impl ErrorKind {
    /// Create an error context from this error
    pub fn context(self, source: impl Into<BoxError>) -> Context<ErrorKind> {
        Context::new(self, Some(source.into()))
    }
}

/// Error type
#[derive(Debug)]
pub struct Error(Box<Context<ErrorKind>>);

impl Deref for Error {
    type Target = Context<ErrorKind>;

    fn deref(&self) -> &Context<ErrorKind> {
        &self.0
    }
}

impl Display for Error {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        self.0.fmt(f)
    }
}

impl std::error::Error for Error {
    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
        self.0.source()
    }
}

impl From<ErrorKind> for Error {
    fn from(kind: ErrorKind) -> Self {
        Context::new(kind, None).into()
    }
}

impl From<Context<ErrorKind>> for Error {
    fn from(context: Context<ErrorKind>) -> Self {
        Error(Box::new(context))
    }
}

impl From<io::Error> for Error {
    fn from(err: io::Error) -> Self {
        ErrorKind::Io.context(err).into()
    }
}
