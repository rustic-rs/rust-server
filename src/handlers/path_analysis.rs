use std::fmt::{Display, Formatter};
use crate::handlers::path_analysis::ArchivePathEnum::{CONFIG, DATA, INDEX, KEYS, LOCKS, NONE, SNAPSHOTS};
use crate::web::{TPE_CONFIG, TPE_DATA, TPE_INDEX, TPE_KEYS, TPE_LOCKS, TPE_SNAPSHOTS, TYPES};
use crate::error::Result;
use crate::error::ErrorKind;

#[derive(Debug, PartialEq)]
pub(crate) enum ArchivePathEnum { DATA, KEYS, LOCKS, SNAPSHOTS, INDEX, CONFIG, NONE }
pub(crate) struct ArchivePath{
    pub(crate) path_type: ArchivePathEnum,
    pub(crate) tpe: String,
    pub(crate) path: String,
    pub(crate) name: String,
}

impl Display for ArchivePath {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "[ArchivePath] path_type = {:?}, path: {}, tpe: {}, name: {:?}",
               self.path_type,
               self.path,
               self.tpe,
               self.name,
        )
    }
}

pub(crate)  fn decompose_path(path:String) -> Result<ArchivePath> {

    // Collect to a list of non empty path elements
    let mut elem:Vec<String> = path
        .split('/')
        .map(|s| s.to_string())
        .filter( |s|{!s.is_empty()})
        .collect();
    let length = elem.len();
    tracing::debug!("[decompose_path] elem = {:?}", &elem);

    let mut ap = ArchivePath{
        path_type: NONE,
        tpe: "".to_string(),
        path: "".to_string(),
        name: "".to_string(),
    };

    if length == 0 {
        tracing::debug!("[decompose_path] Empty path!");
        return Err(ErrorKind::FilenameNotAllowed(path));
    }

    // Analyse tail of the path to find name and type values
    let tmp= elem.pop().unwrap();
    let (tpe, name) = if tmp.eq( TPE_CONFIG ) {
        ap.path_type = CONFIG;
        tracing::debug!("[decompose_path] ends with config");
        if length > 1 {
            let tpe = elem.pop().unwrap();
            if TYPES.contains(&tpe.as_str() ) {
                (tpe, tmp)                      // path = /:path/:tpe/:config
            } else {
                elem.push(tpe);
                ("".to_string(), tmp)           // path = /:path/:config
            }
        } else {
            ("".to_string(), tmp)               // path = /:config
        }
    } else if TYPES.contains(&tmp.as_str()) {
        ap.path_type = get_path_type(&tmp);
        (tmp, "".to_string())                 // path = /:path/:tpe --> but NOT "config"
    } else if length>1 {
        let tpe = elem.pop().unwrap();
        if TYPES.contains(&tpe.as_str()) {
            assert_ne!( tpe.as_str(), TPE_CONFIG ); // not allowed: path = /:path/:config/:name
            ap.path_type = get_path_type(&tpe);
            (tpe, tmp)                        // path = /:path/:tpe/:name
        } else {
            ap.path_type = NONE;
            elem.push(tpe);
            elem.push(tmp);
            ("".to_string(), "".to_string())  // path = /:path --> with length (>1)
        }
    } else {
        ap.path_type = NONE;
        elem.push(tmp);
        ("".to_string(), "".to_string())      // path = /:path --> with length (1)
    };

    ap.tpe = tpe;
    ap.name = name;
    ap.path = elem.join("/");

    tracing::debug!("[decompose_path]  {:}", &ap);

    Ok(ap)
}

fn get_path_type(s:&str) -> ArchivePathEnum {
    match s {
        TPE_CONFIG => {CONFIG},
        TPE_DATA => { DATA  },
        TPE_KEYS  => { KEYS },
        TPE_LOCKS => { LOCKS },
        TPE_SNAPSHOTS => { SNAPSHOTS },
        TPE_INDEX => { INDEX },
        _ => {NONE}
    }
}

#[cfg(test)]
mod test {
    use crate::handlers::path_analysis::decompose_path;
    use crate::web::{TPE_CONFIG, TPE_DATA, TPE_LOCKS};
    use crate::error::Result;
    use crate::handlers::path_analysis::ArchivePathEnum::CONFIG;
    use crate::log::init_tracing;


    #[test]
    fn archive_path_struct() -> Result<()>{
        init_tracing();

        let path = "/a/b/data/name".to_string();
        let ap = decompose_path(path)?;
        assert_eq!( ap.tpe,  TPE_DATA );
        assert_eq!( ap.name,  "name".to_string() );
        assert_eq!( ap.path,  "a/b");

        let path = "/data/name".to_string();
        let ap = decompose_path(path)?;
        assert_eq!( ap.tpe,  TPE_DATA );
        assert_eq!( ap.name,  "name".to_string() );
        assert_eq!( ap.path,  "");

        let path = "/a/b/locks".to_string();
        let ap = decompose_path(path)?;
        assert_eq!( ap.tpe,  TPE_LOCKS );
        assert_eq!( ap.name,  "".to_string() );
        assert_eq!( ap.path,  "a/b");

        let path = "/data".to_string();
        let ap = decompose_path(path)?;
        assert_eq!( ap.tpe,  TPE_DATA );
        assert_eq!( ap.name,  "".to_string() );
        assert_eq!( ap.path,  "");

        let path = "/a/b/data/config".to_string();
        let ap = decompose_path(path)?;
        assert_eq!( ap.path_type,  CONFIG );
        assert_eq!( ap.tpe,  TPE_DATA );
        assert_eq!( ap.name,  "config".to_string() );
        assert_eq!( ap.path,  "a/b");

        let path = "/a/b/config".to_string();
        let ap = decompose_path(path)?;
        assert_eq!( ap.path_type,  CONFIG );
        assert_eq!( ap.tpe,  "".to_string() );
        assert_eq!( ap.name,  "config".to_string() );
        assert_eq!( ap.path,  "a/b");

        let path = "/config".to_string();
        let ap = decompose_path(path)?;
        assert_eq!( ap.path_type,  CONFIG );
        assert_eq!( ap.tpe,  "".to_string() );
        assert_eq!( ap.name,  "config".to_string() );
        assert_eq!( ap.path,  "");

        Ok(())
    }
}