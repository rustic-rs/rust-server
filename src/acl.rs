use once_cell::sync::OnceCell;
use anyhow::Result;
use std::collections::HashMap;
use std::fs;
use std::path::PathBuf;
use serde_derive::Deserialize;
use crate::error::ErrorKind;
use crate::handlers::path_analysis::TPE_LOCKS;

//Static storage of our credentials
pub static ACL:OnceCell<Acl> = OnceCell::new();

pub fn init_acl( state: Acl ) -> Result<(), ErrorKind> {
    if ACL.get().is_none() {
        match ACL.set(state) {
            Ok(_) => {}
            Err(_) => {
                return Err(ErrorKind::InternalError("Can not create ACL struct".to_string()));
            }
        }
    }
    Ok(())
}
// Access Types
#[derive(Debug, Clone, PartialEq, PartialOrd, serde_derive::Deserialize)]
pub enum AccessType {
    Nothing,
    Read,
    Append,
    Modify,
}

// #[derive(Debug, Clone)]
// #[enum_dispatch]
// pub(crate) enum AclCheckerEnum {
//     Acl(Acl),
// }
//
// impl AclCheckerEnum {
//     pub fn acl_from_file(
//         append_only: bool,
//         private_repo: bool,
//         file_path: Option<PathBuf>,
//     ) -> Result<Self> {
//         let acl = Acl::from_file(append_only, private_repo, file_path)?;
//         Ok(AclCheckerEnum::Acl(acl))
//     }
// }

//#[enum_dispatch(AclCheckerEnum)]
pub trait AclChecker: Send + Sync + 'static {
    fn allowed(&self, user: &str, path: &str, tpe: &str, access: AccessType) -> bool;
}

// ACL for a repo
type RepoAcl = HashMap<String, AccessType>;

// Acl holds ACLs for all repos
#[derive(Clone, Deserialize, Debug)]
pub struct Acl {
    repos: HashMap<String, RepoAcl>,
    append_only: bool,
    private_repo: bool,
}

impl Default for Acl {
    fn default() -> Self {
        Self {
            repos: HashMap::new(),
            append_only: true,
            private_repo: true,
        }
    }
}

// read_toml is a helper func that reads the given file in toml
// into a Hashmap mapping each user to the whole passwd line
fn read_toml(file_path: &PathBuf) -> Result<HashMap<String, RepoAcl>> {
    let s = fs::read_to_string(file_path)?;
    // make the contents static in memory
    let s = Box::leak(s.into_boxed_str());

    let mut repos: HashMap<String, RepoAcl> = toml::from_str(s)?;
    // copy key "default" into ""
    if let Some(default) = repos.get("default") {
        let default = default.clone();
        repos.insert("".to_owned(), default);
    }
    Ok(repos)
}

impl Acl {
    pub fn from_file(
        append_only: bool,
        private_repo: bool,
        file_path: Option<PathBuf>,
    ) -> Result<Self> {
        let repos = match file_path {
            Some(file_path) => read_toml(&file_path)?,
            None => HashMap::new(),
        };
        Ok(Self {
            append_only,
            private_repo,
            repos,
        })
    }
}

impl AclChecker for Acl {
    // allowed yields whether these access to {path,tpe, access} is allowed by user
    fn allowed(&self, user: &str, path: &str, tpe: &str, access: AccessType) -> bool {
        // Access to locks is always treated as Read
        let access = if tpe == TPE_LOCKS {
            AccessType::Read
        } else {
            access
        };

        match self.repos.get(path) {
            // We have ACLs for this repo, use them!
            Some(repo_acl) => match repo_acl.get(user) {
                Some(user_access) => user_access >= &access,
                None => false,
            },
            // Use standards defined by flags --private-repo and --append-only
            None => {
                (user == path || !self.private_repo)
                    && (access != AccessType::Modify || !self.append_only)
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use std::env;
    use super::AccessType::*;
    use super::*;

    #[test]
    fn test_static_acl_access() {
        let cwd = env::current_dir().unwrap();
        let acl = PathBuf::new()
            .join(cwd)
            .join("test_data" )
            .join("acl.toml" );

        dbg!(&acl);

        let auth = Acl::from_file(false, true, Some(acl) ).unwrap();
        init_acl(auth).unwrap();

        let acl = ACL.get().unwrap();
        assert!( &acl.private_repo);
        assert!( ! &acl.append_only );
        let access = acl.repos.get("test_repo").unwrap();
        let access_type = access.get("test").unwrap();
        assert_eq!( access_type, &Append );
    }


    #[test]
    fn allowed_flags() {
        let mut acl = Acl {
            repos: HashMap::new(),
            append_only: true,
            private_repo: true,
        };
        assert!(!acl.allowed("bob", "sam", "keys", Read));
        assert!(!acl.allowed("bob", "sam", "data", Read));
        assert!(!acl.allowed("bob", "sam", "data", Append));
        assert!(!acl.allowed("bob", "sam", "data", Modify));
        assert!(!acl.allowed("bob", "bob", "data", Modify));
        assert!(acl.allowed("bob", "bob", "locks", Modify));
        assert!(acl.allowed("bob", "bob", "keys", Append));
        assert!(acl.allowed("bob", "bob", "data", Append));
        assert!(acl.allowed("", "", "data", Append));
        assert!(!acl.allowed("bob", "", "data", Read));

        acl.append_only = false;
        assert!(!acl.allowed("bob", "sam", "data", Modify));
        assert!(acl.allowed("bob", "bob", "data", Modify));

        acl.private_repo = false;
        assert!(acl.allowed("bob", "sam", "data", Modify));
        assert!(acl.allowed("bob", "bob", "data", Modify));
        assert!(acl.allowed("bob", "", "data", Modify));
    }

    #[test]
    fn repo_acl() {
        let mut acl = Acl::default();

        let mut acl_all = HashMap::new();
        acl_all.insert("bob".to_string(), Modify);
        acl_all.insert("sam".to_string(), Append);
        acl_all.insert("paul".to_string(), Read);
        acl.repos.insert("all".to_string(), acl_all);

        let mut acl_bob = HashMap::new();
        acl_bob.insert("bob".to_string(), Modify);
        acl.repos.insert("bob".to_string(), acl_bob);

        let mut acl_sam = HashMap::new();
        acl_sam.insert("sam".to_string(), Append);
        acl_sam.insert("bob".to_string(), Read);
        acl.repos.insert("sam".to_string(), acl_sam);

        // test ACLs for repo all
        assert!(acl.allowed("bob", "all", "keys", Modify));
        assert!(!acl.allowed("sam", "all", "keys", Modify));
        assert!(acl.allowed("sam", "all", "keys", Append));
        assert!(acl.allowed("sam", "all", "locks", Modify));
        assert!(!acl.allowed("paul", "all", "data", Append));
        assert!(acl.allowed("paul", "all", "data", Read));
        assert!(acl.allowed("paul", "all", "locks", Modify));
        assert!(!acl.allowed("attack", "all", "data", Modify));

        // test ACLs for repo bob
        assert!(acl.allowed("bob", "bob", "data", Modify));
        assert!(!acl.allowed("sam", "bob", "data", Read));
        assert!(!acl.allowed("attack", "bob", "locks", Modify));

        // test ACLs for repo sam
        assert!(!acl.allowed("sam", "sam", "data", Modify));
        assert!(acl.allowed("sam", "sam", "data", Append));
        assert!(!acl.allowed("bob", "sam", "keys", Append));
        assert!(acl.allowed("bob", "sam", "keys", Read));
        assert!(!acl.allowed("attack", "sam", "locks", Read));

        // test ACLs for repo paul => fall back to flags
        assert!(!acl.allowed("paul", "paul", "data", Modify));
        assert!(acl.allowed("paul", "paul", "data", Append));
        assert!(!acl.allowed("sam", "paul", "data", Read));
    }
}
