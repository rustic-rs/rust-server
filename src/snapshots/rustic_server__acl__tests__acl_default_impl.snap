---
source: src/acl.rs
expression: acl
---
Acl {
    private_repo: true,
    append_only: true,
    repos: {},
}
